### ISSUE TITLE
Good pattern for updating atoms based on subscription

### ISSUE THREAD

**Issue by tarngerine** [pmndrs/jotai#234]
Lets say I have an API that I can subscribe to, and any time a value changes (can happen once a frame at 60FPS) I would want to update an atom to always reflect that

Currently I can just create a useEffect to do this at the App.tsx level, but wondering if there's an atoms-only way to do this.

some pseudocode:
```
useEffect(() => {
  async function load() {
    setAtom(client.get(someId))
  }
  client.subscribe(someId, (msg) => setAtom(msg));
  () => {
    client.unsubscribe(someId);
  }
}, [])
```

**Comment by dai-shi** [comment]
Jotai is pretty much like `useState` both conceptually and implementation wise.
So, your pseudo code looks pretty much like what it should be.

> wondering if there's an atoms-only way to do this.

I think this falls into recoil's atom effects.
We have some discussions: #211 
(TBH, I'm not sure if/how it fits in jotai, because our atom has no notion of clean up, just rely on GC.)

**Comment by tarngerine** [comment]
Yea I saw the effects! Very cool. Yea not having cleanup would be an issue. Will use effects for now! 😊

**Comment by tarngerine** [comment]
i came up with a slightly different way to model this, but it would require being able to run `set()` from the getter. would this be too weird?

```
const client = atom(client);
const atom1 = atom({});
const atom1Listener = undefined;
const updateAtom1 = atom((get, set) => {
  const client = get(client);
  if (atom1Listener) client.unsubscribe(atom1Listener);
  atom1Listener = client.subscribe("event", (msg) => {
    set(atom1, {msg})
  });
})

```

**Comment by dai-shi** [comment]
We invoke `read` in the render phase. It's a well known discussion that we should avoid side effects in the render phase. Unless it's GC-able, there's no way to clean up.

`set()` in `read` seems too much trouble. People would only misuse it...

I like your way of thinking, and trying to create a good pattern, which I call "Jotai Puzzle."

**Comment by dai-shi** [comment]
Let me try ideating something.

So, what works now is something like this:
```js
const dataAtom = atom(null)

const useDataAtomWithEffect = () => {
  const [data, setData] = useAtom(dataAtom)
  useEffect(() => {
    const init = async () => {
      setData(await fetchData())
    }
    const unsubscribe = subscribeToData(setData)
    init()
    return unsubscribe
  }, [setData])
  return data
}
```

Let's see if we can generalize a bit.

```jsx
// define an atom
const dataAtom = atom(null)
// define an atom initializer
const initializeDataAtom = async (set) => {
  const unsubscribe = subscribeToData(set)
  (async () => { set(await fetchData()) })()
  return unsubscribe
}

// list all atoms with initializer
const atomsWithInitializer = [
  [dataAtom, initializeDataAtom]
]

// internal component
const InitializeAtom = ({ atom, initialize }) => {
  const [value, setValue] = useAtom(atom)
  useEffect(() => {
    return initialize(setValue)
  }, [setValue])
  return null
}

// component to initialize atoms recursively
const InitializeAtoms = ({ atoms, children }) =>
  atoms.reduceRight(a, [atom, initialize] =>
    React.createElement(InitializeAtom, { atom, initialize }, a),
    children,
  )

// How to use
const App = () => (
  <Provider>
    <InitializeAtoms atoms={atomsWithInitializer}>
      <MainLayout />
    </InitializeAtoms>
  </Provider>
)
```

Just a rough idea.

**Comment by nibtime** [comment]
Hi, I am just getting into Jotai. First impression: pretty awesome! 

Then I got curious whether there is a recommended way of reconciling push-based collections (streams) (WebSockets or any kind of PubSub protocol, browser events, web workers) with React state, i.e. expressing this with the atom abstraction of Jotai. As I see there are already some ideas on this.

The nicest way I've seen for this up to this point (but I am not really a fan Rx for various reasons - just so many things you can do wrong 😃)  is the [`useObservableSuspense`](https://observable-hooks.js.org/guide/render-as-you-fetch-suspense.html#benefits-of-observable-as-data-source)  hook ([sandbox sample](https://codesandbox.io/s/github/crimx/observable-hooks/tree/master/examples/suspense)) which essentially enables you to map the stream to React Nodes and render it in Suspense, just as you would with a pull-based collection (i.e. array). The good thing Rx offers though is, that you can compose streams like collections (dangerous though, because operators can alter stream semantics) and especially the composition of patterns that involve time.

Would it be possible to build a similar kind of abstraction with Jotai? Or is something like that beyond the scope of Jotai?

**Comment by dai-shi** [comment]
@nibtime This is something we want to support on top of jotai core. `jotai/query` is the first attempt, and that required to implement `onMount` in core. Check out src/query/atomWithQuery.ts. It's basically using an observable in RQ. The implementation is a bit tricky. Would be nice if we have some ideas to improve it (without affecting much to core. preferably zero.)

Having written this, I noticed I might have misunderstood your point. Is your question about if we can do something like rx operators on atoms, rather than using existing rxjs? I haven't thought about it. Would be interesting to explore if it could be done on top of jotai core.

**Comment by nibtime** [comment]
@dai-shi No, you didn't misunderstand my point at all. I wanted to just throw in an idea on what I think is conceptually the right direction for dealing with this "push" data source problem. I don't have any assumption on how this is best to be implemented. Could be a package `jotai/rx` on top of `rx` that reconciles observables with Jotai (so essentially what the `use-observable` package does, but with atoms). But having a first-class stream abstraction in Jotai without extra dependencies would be superb too, but surely a lot more work than reusing `rx`. Also, using Jotai and `use-observable` side by side could be fine too, since it is all just React hooks (so there would just be some sort of docs that display a scenario using them side-by-side and a clear recommendation when this makes sense). In the end, it's just about using the right thing for a given task, I believe.

I feel when it comes to "push" data sources with React, there's some sort of vacuum and then patterns are usually hacked away with `useEffect`, imperative logic in event handlers, and timers over and over again. No existing state manager for React touches them conceptually (except `use-observable`, but I don't see it as a state manager, it's more a "state reconciler").

I just looked a little bit into `react-query` and `jotai/query`. I didn't know `react-query` before, but I used `swr` before, so it looked quite familiar. That is absolutely great too, I would most certainly rather use `jotai/query` than `swr` given it's stable. 

But I think it covers a different, but related and also very important use case as far as I can tell. And the notion of a stream is somehow also inherent in React (the stream of React state and re-renders). This makes it very confusing for me to classify things the right way. My mental models are as follows: 

**`jotai/query`/`react-query`** "I need real-time data from repeatedly polling/re-fetching a request-response ("pull") data source. Please turn it into a stream of React state (and handle all the very annoying things like caching, caching, and caching 😄)" - but I really don't know `react-query` and it seems to be quite comprehensive. However, in the doc search, there are 0 hits on "WebSocket", so I assume it doesn't cover "push" data sources.

**`use-observable`/`rx`** "I have a data source that emits real-time data over a "push" protocol, please turn it into a stream of React state or a stream of real-time UI updates (Suspense)." - The concept of caching makes no sense here because the data source proactively pushes changed data only. But of course, problems just don't magically disappear, they just get replaced with something else: backpressure 😄. 

As I mentioned, `use-observable` and `rx` are the only really good way I know how to deal with "push" in React. I am trying to keep it to a minimum though and only use it in cases where it really makes sense. I will lay out some of my experiences and use cases for `rx`, and hopefully, this can help a little in finding the right abstractions for Jotai or to determine what the role of Jotai in such scenarios could be. I do not have deep knowledge of Jotai and its abstractions yet, but would be very interested to get involved in the project!

<details>
<summary>my rules for use-observable</summary>

####
1. there are only 3 allowed operators to source a stream: [`fromEventPattern`](https://rxjs-dev.firebaseapp.com/api/index/function/fromEventPattern) (turn an arbitrary API with a subscribe/unsubscribe pattern into an observable stream). [`fromEvent`](https://rxjs-dev.firebaseapp.com/api/index/function/fromEvent) (convenience function for browser events assuming `addEventHandler` and `removeEventHandler`) and [`webSocket`](https://rxjs-dev.firebaseapp.com/api/webSocket/webSocket) (return a multicast subject that streams data updates from the WebSocket endpoint and that can be used to send messages to the endpoint as well)
2. I need to do transformations that would be really cumbersome and awkward without `rx`
3. leave `rx` / reconcile with React as soon as possible

</details>

<details>
<summary>most important compositions</summary>

#### [Overview of operators](https://rxjs-dev.firebaseapp.com/api)

#### basic compositions 
stuff like map, variants of chain (flatMap, switchMap, exhaustMap, combineMap), scan, ...

I realize though, that this might be a bit too opinionated (FP) for Jotai. 

#### `debounceTime`
emit most recent data update only after a given time interval of "emission silence".

use-case:  any sort of "fetch-as-you-type" scenario: async form validation, fetching autocomplete data, ...

#### `sampleTime` 
make the data stream "less precise". Emit only the most recent data update per given time interval. Of course, if sampling is offered via a protocol that always has to be preferred.

use-case: the easiest way to deal with backpressure. E.g. you want to display sensory data in UI, but the sensor emits data way faster than you can re-render (e.g. binocular eye-tracker with 200hz)  

#### `pairwise` / `bufferCount` / `bufferTime`
group data emissions together, either by count or time.

use-case: Whenever you want to do some kind of "averaging" over the stream. E.g. you want to determine scroll direction from `scroll` events.

```ts
const scrollPositionObservable = fromEvent(window, "scroll", { passive: true }, (event) => event.target.scrollingElement)
.pipe(map((target) => (target as any).scrollTop as number), pairwise())

const [previousScrollPosition, scrollPosition] = useObservableState(scrollPositionObservable, [0, 0])
const isScrollingUp = scrollPosition < previousScrollPosition
const isScrollingDown = scrollPosition > previousScrollPosition
```

</details>

<details>
<summary>singlecast (cold) vs. multicast (hot) semantics</summary>

####
To determine whether a data stream is single-cast or multicast I just ask the following question:

> "Do data updates make sense without a baseline?"

If yes, then the stream can be multicast, if no, then it has to be single-cast (but it could be turned into multicast, e.g. by merging in diffs with `scan`, so it's emitting "absolute" values.

Example for single-cast: [Firestore collection real-time updates](https://firebase.google.com/docs/firestore/query-data/listen). On subscription, you receive the whole collection (can be filtered of course) as `docChanges` of type `added` and then only diffs/deltas of type `added`, `modified`, or `removed` are pushed to you. Those make no sense without the initial baseline because data updates are "relative". But also "absolute" values, i.e. the whole changed collection is pushed, so those changes could be treated with multicast semantics. Another example I want to start playing around with a little is the [WebSocket API of BitMEX](https://www.bitmex.com/app/wsAPI). It's the same here, you receive the initial data structure upon subscription and then diffs/deltas (called "partials") are pushed via WebSocket.

Example for multicast: Browser events like scroll, mouse position, etc. You can just chime in at any point of the stream execution and do meaningful stuff because data updates are "absolute"
</details>

**Comment by dai-shi** [comment]
@nibtime

> I do not have deep knowledge of Jotai and its abstractions yet, but would be very interested to get involved in the project!

That would be great! I could naively implement `atomWithObservable`, but rather wait for you to chime in. You seem to have far better understanding around this topic.

**Comment by nibtime** [comment]
@dai-shi 
it probably really only seems so 😄. I have an understanding only as a consumer of things, not as a producer, which requires deeper internal knowledge of React I don't have (which I hope I can improve by getting involved).

So as an idea on how I could initially contribute something: I could try to cook up some sample that involves a "push" data source. I am not a friend of mocking and would prefer it to be something that streams for real. I just thought of a very simple market depth chart for a BitMEX instrument that updates in real-time. The BitMEX API has public WebSocket channels, so that would be good for sharing with other people. I would try to implement it both with `use-observable` and Jotai for a side-by-side comparison. So basically a fun experiment. Would something like that help?

As for the producer part:  In implementing `atomWithObservable` I can't be of much help at this point. You surely know everything about Jotai and have a deep understanding of React. I'd like to learn from that first and try to assist in developing the right ideas for this.
 
If you would try to implement `atomWithObservable` I suspect that you have to deal with similar problems like `use-observable`. The author of that package has done a tremendous job (docs are great too). I wouldn't dare using React with `rx` otherwise. So it could be worthwhile looking into it or even get in touch with the author somehow (@crimx).

**Comment by dai-shi** [comment]
@nibtime OK, got it. Here's the deal. I will quickly draft `atomWithObservable`, and you can help all the rest, experimenting, reporting/fixing bugs, making examples, writing docs and tutorials would be good. tests would also be important. And if all seem good, we can release it.

**Comment by nibtime** [comment]
@dai-shi All right. I will set up a Next project for experimentation and start with the Jotai implementation when you drafted `atomWithObservable`.

**Comment by ricmello** [comment]
Hey guys, what's the current status of this feature? Do you need any help? 

I tried to test in a sample project but without success. Here is the sample link:
https://stackblitz.com/edit/react-ts-fnnrni?file=index.tsx

The error thrown is `Cannot read properties of undefined (reading 'subscribe')`.

I tried to pass a Subject, a BehaviourSubject and an Observable but none worked.

**Comment by dai-shi** [comment]
`atomWithObservable` should be fully working now. Your help is appreciated to test it and improve it (especially, docs and tutorial, and of course finding bugs.)

It requires `<Suspense>`:
https://stackblitz.com/edit/react-ts-2cmvbz?file=index.tsx

> The error thrown is `Cannot read properties of undefined (reading 'subscribe')`.

I don't see this error. 🤔

**Comment by ricmello** [comment]
Cool! I'll definitely take a look on it.

In which browser did you tested? I just tested on Safari and it worked so I thought I was going crazy, but now I realized that the problem only happens on Chrome. Both Safari and Firefox works well.

**Comment by dai-shi** [comment]
I tested on Chrome... 😅

**Comment by ricmello** [comment]
Oh man! I'm not crazy I swear 😆 

<img width="1432" alt="Screen Shot 2021-10-04 at 19 38 56" src="https://user-images.githubusercontent.com/4468451/135934038-7a23133a-98ef-4295-ad9e-0e921d300b83.png">

It doesn't make any sense to me, but now I disabled Chrome's Redux DevTools extension and it worked on Chrome.

**Comment by dai-shi** [comment]
Confirmed the error with Redux DevTools Extension. Weird. Could be a stackblitz issue?

**Comment by ricmello** [comment]
I don't think so because the first time that I saw it was testing in my local app. I'll debug it tomorrow.

**Comment by dai-shi** [comment]
Oh, really. Looking forward to your debug. Thanks for your interest.

**Comment by neckaros** [comment]
What do you think about a version of this utility without Promise but with an initial value provided? For people who don’t want to use experimental Suspense yet?
At least that the only use of promise I understood from the code for this utility

**Comment by dai-shi** [comment]
`atomWithObservable` with initial value to avoid suspense? That's possible. We did something like that in `atomWithQuery` in `jotai/query`. Would you like to open a new issue for tracking?

> For people who don’t want to use experimental Suspense yet?

For this specific purpose, we have `loadable`.
